### DISCUSSION TITLE
Split deposit in multiple streams

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/v2-periphery#4] [Ideas]
A user in our Discord server asked for this feature. Their use case was a DAO context - instead of having to create multiple proposals, they would prefer to create just one, and provide one deposit amount, and then Sablier to take that and create multiple streams.

<img width="1331" alt="Screen Shot 2022-07-20 at 12 56 18 AM" src="https://user-images.githubusercontent.com/8782666/179855812-a4a75105-7bc1-4da7-87fc-62ace5dbd8d8.png">

We should very easily be able to do this via a [PRBProxy](https://github.com/paulrberg/prb-proxy)-enabled target contract, or potentially some dedicated router in the periphery so that DAOs don't have to deploy a PRBProxy.

Also relevant: [0xsplits.xyz](https://www.0xsplits.xyz/). Cc @razgraf.

**Comment by andreivladbrg** [comment]
>We should very easily be able to do this via a [PRBProxy](https://github.com/paulrberg/prb-proxy)-enabled target contract, or potentially some dedicated router in the periphery so that DAOs don't have to deploy a PRBProxy.

I think you over complicate this way. 

Would be much easier to create an internal mapping of `balances` and add a different function for `deposit`.

There will be 3 different functions:
1. create
2. deposit
3. createAndDeposit

Actually llamapay has also done that: https://github.com/LlamaPay/llamapay/blob/90d18e11b94b02208100b3ac8756955b1b726d37/contracts/LlamaPay.sol#L217-L226

**Reply by PaulRBerg** [reply]
Again, the goal is to keep the core contracts as minimal as possible. We might "over complicate" things in the periphery sometimes, but that's fine (though I don't even this case is an "over complication", splitting a deposit amount is easy).

> Would be much easier to create an internal mapping of `balances` and add a different function for deposit.

This would be a major change, which would open up a can of worms.

**Comment by razgraf** [comment]
Wait. Maybe I'm not understanding the use-case but... Isn't this what the batching method (that was supposed to live either in our periphery or in a proxy) would do? Create "multiple streams" in the same transaction/deposit?

@andreivladbrg 
> Actually llamapay has also done that

I think this is just a side effect of their no upfront deposit mechanism. They create a virtual balance of the depositor which is then split between the streams. So they need this split between deposit and create to allow for top-ups.

**Reply by PaulRBerg** [reply]
Nope.

Batching creates multiple streams given multiple deposit amounts.

Splitting creates multiple streams given one deposit amount.

**Reply by razgraf** [reply]
The first one still requires only one transaction to be coded into the gouvernance queue. It looks like a preference if you think of it as "depositing 1000 that's split between 10 people" versus "depositing 100 dollars for 10 streams for a total deposit of 1000". Again am I missing something?

**Reply by PaulRBerg** [reply]
Yeah, you're right, splitting is a special case of batching.

I think we should be able to keep only 1 batch-create implementation, and the UI would split the deposit amount equally between all recipients.

**Comment by PaulRBerg** [comment]
Locking based on the realization above that splitting is a special case of batching.
